pkgs/development/compilers/julia/build-julia-package.nix

```diff
@@ -0,0 +1,163 @@
+, buildInputs ? []
+
+# Propagate build dependencies so in case we have A -> B -> C,
+# C can import package A propagated by B
+# Run-time dependencies for the package.
+, propagatedBuildInputs ? []
+
+# Julia packages that are required at runtime for this one.
+# These behave similarly to propagatedBuildInputs, where if
+# package A is needed by B, and C needs B, then C also requires A.
+# The main difference between these and propagatedBuildInputs is
+# during the package's installation into Julia, where all
+# requiredJuliaPackages are ALSO installed into Julia.
+, requiredJuliaPackages ? []
+
+, preBuild ? ""
+, meta ? {}
+, passthru ? {}
+
+# Julia packages can distribute 'Artifacts'.  These can be some data,
+# binary, ...  We Package 'Artifacts' as independent derivations. They
+# differ from the rest of the packages in that they don't get compiled
+# and the package providing it expect to find them in a specific location.
+, isJuliaArtifact ? false
+
+, ... } @ attrs:
+
+let
+  requiredJuliaPackages' = computeRequiredJuliaPackages requiredJuliaPackages;
+  juliaDepotPath = computeJuliaDepotPath requiredJuliaPackages';
+  juliaLoadPath = computeJuliaLoadPath requiredJuliaPackages';
+  juliaArtifacts = computeJuliaArtifacts requiredJuliaPackages';
+
+  # This step is required because when
+  # a = { test = [ "a" "b" ]; }; b = { test = [ "c" "d" ]; };
+  # (a // b).test = [ "c" "d" ];
+  # This used to mean that if a package defined extra nativeBuildInputs, it
+  # would override the ones for building a Julia package (Julia itself)
+  # causing everything to fail.
+  attrs' = builtins.removeAttrs attrs [
+    "nativeBuildInputs"
+    "buildInputs"
+    "requiredJuliaPackages"
+  ];
+
+in stdenv.mkDerivation ({
+  packageName = "${fullPkgName}";
+  # The name of the package ends up being
+  # "julia-version-package-version"
+  name = "${julia.pname}-${julia.version}-${fullPkgName}";
+
+  # This states that any package built with the function that this returns
+  # will be a julia package. This is used for ensuring other julia
+  # packages are installed into julia during the environment building phase.
+  isJuliaPackage = true;
+
+  requiredJuliaPackages = requiredJuliaPackages';
+
+  nativeBuildInputs = nativeBuildInputs ++ [ julia patchelf ];
+
+  buildInputs = buildInputs ++ requiredJuliaPackages' ++ [ stdenv.cc.libc ];
+
+  propagatedBuildInputs = propagatedBuildInputs;
+
+  preUnpack = ''
+    mkdir ${attrs.pname}
+    cd ${attrs.pname}
+  '';
+
+  setSourceRoot = ''
+    sourceRoot=$PWD
+  '';
+
+  # Set NIX_JULIA_PRECMD to the currntly only known command needed so
+  # as to maximize the number of packages that compile without extra
+  # configuration (see https://github.com/JuliaGraphics/Gtk.jl/issues/346)
+  configurePhase = ''
+    runHook preConfigure
+    export NIX_JULIA_PRECMD=${xvfb-run}/bin/xvfb-run
+    runHook postConfigure
+  '';
```

Comment:
Why do we export this as a bash variable? This should be a nix string which we could move into a let in to not duplicate it

Reply:
The idea is that if the need arises, a user can override it in a `preBuild` phase.